### PR TITLE
Support author protocols rule

### PR DESCRIPTION
**Pull request by diehuxx**
Addresses #241. `recipient` is already implemented. This PR adds support for `author`

### CODE REVIEW

File: src/core/protocol-authorization.ts
Comment (thehenrytsai) [inline src/core/protocol-authorization.ts:161]: For my understanding: what's the reason that prompted the decision to merge the logic into `verifyAllowedActions()`? Was it because it's hard to understand? Or redundant code? Or?

The intention behind having two separate methods here were:
1. It is easier to know what the method is doing (in theory) without reading the implementation.
2. It gives precise error message when error does occur.

The new implementation removes both benefits above, hence want to discuss it.

File: json-schemas/protocol-rule-set.json
Comment (thehenrytsai) [inline json-schemas/protocol-rule-set.json:55]: Consider filing a good-first-issue for the community to refactor the repeated definition into its own schema so we can just reference the same schema.

**Review by thehenrytsai — COMMENTED** [review]
Thanks @diehuxx, looks good overall!

Have a design question for discussion. Will wait for response before reviewing further!

File: tests/interfaces/records/handlers/records-write.spec.ts
Comment (diehuxx) [reply tests/interfaces/records/handlers/records-write.spec.ts:5]: Yes, I was being a goofball since it's just a test fixture. I'm down to correct the misspelling if it's confusing/annoying/too silly.

File: src/core/protocol-authorization.ts
Comment (diehuxx) [reply src/core/protocol-authorization.ts:161]: I tried keeping `verifyAllowedRequester` and `verifyAllowedActions` separate, but I found that they were becoming redundant. Plus, I felt that the implementation was more intuitive by combining them. I'm down to jam on keeping them separate if you feel that's best. Here was my thought process:

We had these lines originally:
```typescript
    if (allowRule.anyone !== undefined) {
      allowedActions = allowRule.anyone.to;
    } else if (allowRule.recipient !== undefined) {
      allowedActions = allowRule.recipient.to;
    }
    const inboundMessageAction = methodToAllowedActionMap[incomingMessageMethod];
    if (!allowedActions.includes(inboundMessageAction)) {
      throw new Error(`inbound message action '${inboundMessageAction}' not in list of allowed actions ${allowedActions}`);
    }
```

Consider an `allowRule` with both `anyone` and `recipients` defined. It's possible that `recipients` have more allowed actions than `anyone`, and that the `inboundMessageAction` is in `recipient.to`. The above code would erroneously block authorization because it would hit the `if (allowRule.anyone !== undefined)`.

We could fix this by turning the `else if` into a regular `if`, and adding allowed rules into a set of `allowedActions`. But it gets odd when you add `author`.
```typescript
    if (allowRule.anyone !== undefined) {
      allowRule.anyone.to.forEach(action => allowedActions.add(action));
    }
    if (allowRule.recipient !== undefined) {
      allowRule.recipient.to.forEach(action => allowedActions.add(action));
    }
    if (allowRule.author !== undefined) {
      allowRule.author.to.forEach(action => allowedActions.add(action));
    }
```

This is bad because we don't want to allow all the actions in all the rules. We need to check that the message is actually from an `author` or a `recipient`.

```typescript
    if (allowRule.anyone !== undefined) {
      allowRule.anyone.to.forEach(action => allowedActions.add(action));
    }
    if (allowRule.recipient !== undefined) {
      if message is from a recipient {
        allowRule.recipient.to.forEach(action => allowedActions.add(action));
      }
    }
    if (allowRule.author !== undefined) {
      if message is from an author {
        allowRule.recipient.to.forEach(action => allowedActions.add(action));
      }
    }
```
We already do those checks in `verifyAllowedRequester`, so `verifyAllowedRequester` is redundant.

File: json-schemas/protocol-rule-set.json
Comment (diehuxx) [reply json-schemas/protocol-rule-set.json:55]: Good call. #286

File: src/core/protocol-authorization.ts
Comment (thehenrytsai) [reply src/core/protocol-authorization.ts:161]: Thanks @diehuxx for the full context, I did not at all consider the possibility of having multiple "roles" under the same `allow` property! I will continue with the review with this new knowledge!

File: tests/interfaces/records/handlers/records-write.spec.ts
Comment (thehenrytsai) [reply tests/interfaces/records/handlers/records-write.spec.ts:5]: I have spell checker on so the one dozen occurrences stick out like a sore thumb, but knowing this is intentional I'll let you decide if you want to keep the fun or not!

Don't be surprised if someone in the community submits a PR to fix the "typos"!

File: tests/interfaces/records/handlers/records-write.spec.ts
Comment (thehenrytsai) [inline tests/interfaces/records/handlers/records-write.spec.ts:580]: The test is not valid because Alice is writing to her own DWN so even without the `author` rule, the test will pass.

So you'll need have to involve Bob somehow in the scenario.

File: tests/interfaces/records/handlers/records-write.spec.ts
Comment (thehenrytsai) [inline tests/interfaces/records/handlers/records-write.spec.ts:907]: Consider merge this the negative test with the positive test above, tagging on the imposter test at the end, as the test is so verbosely similar.

Comment (diehuxx) [reply tests/interfaces/records/handlers/records-write.spec.ts]: Merged

File: tests/interfaces/records/handlers/records-write.spec.ts
Comment (diehuxx) [reply tests/interfaces/records/handlers/records-write.spec.ts:581]: Fixed

File: tests/interfaces/records/handlers/records-write.spec.ts
Comment (diehuxx) [reply tests/interfaces/records/handlers/records-write.spec.ts]: I switched to the correct spelling. I've had my fun, that's enough for today :)

**Review by thehenrytsai — APPROVED** [review]
Good to go!!! 🙏 🎖️

**Pull request merged**
